vividus-plugin-web-app-to-rest-api/src/main/java/org/vividus/ssllabs/SslLabsClient.java

```diff
@@ -0,0 +1,131 @@
+import java.net.URI;
+import java.time.Duration;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.hc.core5.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.http.client.HttpResponse;
+import org.vividus.http.client.IHttpClient;
+import org.vividus.util.json.JsonUtils;
+import org.vividus.util.wait.DurationBasedWaiter;
+
+public class SslLabsClient
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(SslLabsClient.class);
+    private static final String SSL_SCAN_FAILURE =
+            "SSL scan has not been performed successfully during specified waiting period";
+    private static final String ERROR_MESSAGE = "Status message '{}' received for host {}";
+    private static final String API_VERSION = "/api/v3";
+    private static final String ANALYZE_CALL = "/analyze?host=%s&fromCache=on&maxAge=1";
+    private static final int TOO_MANY_REQUESTS = 429;
+    private static final int SERVICE_IS_OVERLOADED = 529;
+    private static final DurationBasedWaiter WAITER = new DurationBasedWaiter(Duration.ofMinutes(10),
+            Duration.ofSeconds(30));
+
+    private final IHttpClient httpClient;
+    private final JsonUtils jsonUtils;
+    private final String sslLabHost;
+
+    public SslLabsClient(IHttpClient httpClient, JsonUtils jsonUtils, String sslLabHost)
+    {
+        this.httpClient = httpClient;
+        this.jsonUtils = jsonUtils;
+        this.sslLabHost = sslLabHost;
+    }
+
+    public Optional<Grade> performSslScan(String host)
+    {
+        return waitReadyStatus(host).map(AnalyzeResponse::getEndpoints).stream()
+                .flatMap(List::stream)
+                .map(Endpoint::getGrade)
+                .map(Grade::fromString)
+                .min(Comparator.comparingInt(Grade::getGradeValue));
+    }
+
+    private Optional<AnalyzeResponse> waitReadyStatus(String host)
+    {
+        Optional<AnalyzeResponse> response = WAITER.wait(() -> analyze(host).map(r ->
+        {
+            if (Status.ERROR.name().equals(r.getStatus()))
+            {
+                LOGGER.atError().addArgument(r.getStatusMessage()).addArgument(r.getHost()).log(ERROR_MESSAGE);
+                return null;
+            }
+            return r;
+        }), r -> r.isEmpty() || r.get().getStatus().equals(Status.READY.name()));
+
+        if (response.isPresent() && response.get().getStatus().equals(Status.READY.name()))
+        {
+            return response;
+        }
+        LOGGER.atError().log(SSL_SCAN_FAILURE);
+        return Optional.empty();
+    }
+
+    private Optional<AnalyzeResponse> analyze(String host)
+    {
+        Optional<HttpResponse> sslLabsResponse = WAITER.wait(() ->
+        {
+            try
+            {
+                HttpResponse response = httpClient.doHttpGet(URI.create(
+                        String.format("%s%s%s", sslLabHost, API_VERSION, String.format(ANALYZE_CALL, host))));
```

Comment:
what if I specify https://api.ssllabs.com/ ?

Reply:
> what if I specify https://api.ssllabs.com/ ?

`Caused by: java.lang.IllegalArgumentException: URI path begins with multiple slashes`

But I'm not sure that users will change the host too often to add processing for such cases